### ISSUE TITLE
Add support for custom output directory different from the web root

### ISSUE THREAD

**Issue by matteocontrini** [Eptagone/Vite.AspNetCore#21]
A possible use case for this library is to have the dist folder being output to `wwwroot/dist`, or similar.

This makes it possible to put other static assets in `wwwroot` (outside `dist`), which would otherwise be cleared at every Vite build.

I know we could use the Vite `public` folder, but that doesn't work well with the IDE IntelliSense and I think it makes it impossible to use tag helpers while in development (since the `public` folder is not there before a prod build).

The problem is that at the moment `Vite.AspNetCore` expects the `manifest.json` and other output assets to be at the root of the web root, regardless of what the `outDir` in the Vite configuration says.

There are workarounds (see [my blog post](https://matteosonoio.it/aspnet-core-vite/) where I override the web root file provider), but it would be great if the library allowed to override the Vite output directory it expects, while still defaulting to the web root.

**Comment by Eptagone** [comment]
Hi @matteocontrini, you are right. The library should be able to handle those cases.
That feature will be added and included with the next version.

**Comment by wesdevpro** [comment]
> There are workarounds (see [my blog post](https://matteosonoio.it/aspnet-core-vite/) where I override the web root file provider), but it would be great if the library allowed to override the Vite output directory it expects, while still defaulting to the web root.

Thanks for your blog post! I ran into a few problems the first one was that npm was complaining about vue being installed which i easily installed. And then the ASP.VOre WebAPP is complaining about all the .ts files in the node_modules. folder any suggestion?

@matteocontrini

**Comment by Eptagone** [comment]
Hi @matteocontrini. This issue has been fixed with the commit f37bb5cd56bb93ad9f64b1d92c1ed34cba91ab94.
Now, set the `Vite:Base` option to change the output subfolder of your assets. For example, if you want your output to be `wwwroot/dist`, set your options like this:

```JSONC
// asppsettings.json
{
    // ...
    "Vite": {
        // ...
        "Base": "/dist/"
    }
}
```

```TypeScript
const config: UserConfig = {
    // ...
    base: '/dist/',
}
```

The `Vite:Base` value should be the same as your vite [base](https://vitejs.dev/config/shared-options.html#base) option.

**Comment by matteocontrini** [comment]
@Eptagone thank you!

A couple of things:

- the readme says that the `Base` is relative to the content root path but I believe you mean the web root, right?
- to make things work I had to move the dev middleware before the `.UseStaticFiles()` call, otherwise the `wwwroot/dist` takes priority and is incorrectly served even during development (which can happen if I leave there the dist files from another build). In a previous issue you [mentioned](https://github.com/Eptagone/Vite.AspNetCore/issues/12#issuecomment-1530880127) that the dev middleware should be the last one, is there a particular reason for that?

Thanks.

P.S. you might want to use the `jsonc` format (instead of `json`) for markdown code blocks in the readme, so that comments are recognized as valid.

**Comment by Eptagone** [comment]
Yes, it's the web root. I made a little mistake with the docs.

StaticAssets is a middleware too. You have to place the middlewares in the desired order. The first to generate a response to the request is the one that will have priority over the others.

**Comment by matteocontrini** [comment]
Yes, I understand, but wasn't the problem with routes being overriden by Vite fixed with the commit linked in that comment? In that case, there shouldn't be any disadvantage in having the Vite middleware at the top, right?

---

Another thing if I may: when using a custom `Base` path, how should I import the Vite assets in the views?

Both of these seem to work:

```
<link rel="stylesheet" vite-href="~/main.ts" />
<script type="module" vite-src="~/main.ts"></script>
```

```
<link rel="stylesheet" vite-href="~/dist/main.ts" />
<script type="module" vite-src="~/dist/main.ts"></script>
```

And I'm not even sure how's that possible 🧐 the first one produces a request to `/main.ts` which is not exposed by the dev server (it needs to be `/dist/main.ts`), but it works anyway somehow? It's late here so I might be missing something obvious 😄 

Thanks!

**Comment by Eptagone** [comment]
The commit in the other issue was to make the vite middleware respect the order of other middlewares. The order of the middlewares should be defined by you, depending on the needs of your project.
And about the paths, I also noticed that they both work. I think it's because of the @vite/client script. But only the second option should work in production, so use the second one.
